code/datums/extensions/exonet.dm

```diff
@@ -0,0 +1,161 @@
+			return disconnect_result // There was a problem.
+
+	var/datum/exonet/exonet = GLOB.exonets[new_ennid]
+	if(!exonet)
+		return "Error encountered when trying to register \the [holder] to the '[new_ennid]' network."
+	else
+		if(!exonet.router)
+			return "Error encountered when trying to register \the [holder] to the '[new_ennid]' network."
+		if(exonet.router.keydata != keydata)
+			return "Invalid key. Unable to connect to network."
+		exonet.add_device(holder, keydata)
+		ennid = new_ennid
+		netspeed = nic_netspeed
+	return FALSE // This is a success.
+
+/datum/extension/exonet_device/proc/disconnect_network(var/mob/user)
+	if(!ennid)
+		return
+
+	var/datum/exonet/old_exonet = GLOB.exonets[ennid]
+	if(old_exonet)
+		old_exonet.remove_device(holder)
+		ennid = null
+	return FALSE // This is a success.
+
+/datum/extension/exonet_device/proc/broadcast_network(var/b_ennid, var/key)
+	// Broadcasts an ENNID. If the network doesn't exist, it will create it.
+	// If the network does exist, this will attempt to be added as a relay.
+	var/datum/exonet/exonet = GLOB.exonets[b_ennid]
+	if(!exonet)
+		exonet = new(b_ennid)
+		exonet.set_router(holder)
+	else if(exonet.router.keydata != key)
+		return // Security fail.	
+	exonet.add_device(holder)
+	ennid = b_ennid
+	netspeed = NETWORKSPEED_ETHERNET
+
+/datum/extension/exonet_device/proc/get_nearby_networks(var/nic_netspeed)
+	var/list/results = list()
+	for(var/ennid in GLOB.exonets)
+		var/datum/exonet/exonet = GLOB.exonets[ennid]
+		if(exonet.get_signal_strength(holder, nic_netspeed) > 0)
+			results |= exonet
+	return results
+
+/datum/extension/exonet_device/proc/get_local_network()
+	var/datum/exonet/network = GLOB.exonets[ennid]
+	if(!network)
+		return
+	if(network.get_signal_strength(holder, netspeed) > 0)
+		return network
+
+/datum/extension/exonet_device/proc/get_all_networks()
+	return GLOB.exonets
+
+/datum/extension/exonet_device/proc/get_mainframes()
+	var/datum/exonet/network = get_local_network()
+	if(!network)
+		return
+	return network.mainframes
+
+/datum/extension/exonet_device/proc/get_network_tag(var/obj/device)
+	// Gets a friendly, unique name for a device on a local network.
+	var/datum/exonet/network = get_local_network()
+	if(!network)
+		return
+	if(istype(device, /obj/machinery/computer/exonet))
+		var/obj/machinery/computer/exonet/exonet_machine = device
+		if(exonet_machine.net_tag)
+			return exonet_machine.net_tag
+	var/index = network.network_devices.Find(device)
+	return replacetext("[device.name].[index]", " ", "_")
+
+/datum/extension/exonet_device/proc/get_device_by_tag(var/net_tag)
+	if(!net_tag)
+		return
+	var/datum/exonet/network = get_local_network()
+	if(!network)
+		return
+	for(var/obj/machinery/computer/exonet/exonet_machine in network.network_devices)
+		if(!exonet_machine.net_tag)
+			continue
+		if(exonet_machine.net_tag == net_tag)
+			return exonet_machine
+	var/list/tokens = splittext(net_tag, ".")
+	var/index = text2num(tokens[length(tokens)])
```

Comment:
that looks fragile as fuck, unless every time a device is removed form network you loop over every single device and update their network tag to update their indext in the list.

Reply:
Also it seems these can be changed by user, who will probably fail to set right list index.